engine/src/main/java/org/terasology/rendering/nui/widgets/types/builtin/util/FieldsWidgetBuilder.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    public FieldsWidgetBuilder(TypeInfo<T> type, TypeWidgetLibrary library,
+                               Predicate<Field> fieldPredicate) {
+        this.type = type;
+        this.library = library;
+
+        for (Field field : ReflectionUtils.getAllFields(type.getRawType())) {
+            if (Modifier.isTransient(field.getModifiers()) || Modifier.isStatic(field.getModifiers())) {
+                continue;
+            }
+
+            if (!fieldPredicate.test(field)) {
+                continue;
+            }
+
+            Type resolvedFieldType = ReflectionUtil.resolveType(type.getType(), field.getGenericType());
+
+            fields.put(field, TypeInfo.of(resolvedFieldType));
+        }
+    }
+
+    public static <F> Optional<Binding<F>> getFieldBinding(Binding<?> binding, Field field) {
+        if (Modifier.isPublic(field.getModifiers())) {
+            return Optional.of(
+                getAccessibleFieldBinding(binding, field)
+            );
+        }
+
+        return getPropertyBinding(binding, field);
+    }
+
+    private static <F> Optional<Binding<F>> getPropertyBinding(Binding<?> binding, Field field) {
+        Method setter = ReflectionUtil.findSetter(field);
+
+        if (setter == null) {
+            return Optional.empty();
+        }
+
+        Method getter = ReflectionUtil.findGetter(field);
+
+        if (getter == null) {
+            LOGGER.error("Cannot bind field {} with setter {} but no getter", field, setter);
+            return Optional.empty();
+        }
+
+        Binding<F> fieldBinding = new Binding<F>() {
+            @Override
+            public F get() {
+                try {
+                    return (F) getter.invoke(binding.get());
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Unreachable");
+                } catch (InvocationTargetException e) {
+                    throw new RuntimeException(e.getCause());
+                }
+            }
+
+            @Override
+            public void set(F value) {
+                try {
+                    setter.invoke(binding.get(), value);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Unreachable");
+                } catch (InvocationTargetException e) {
+                    throw new RuntimeException(e.getCause());
+                }
+            }
+        };
+
+        return Optional.of(binding.makeChildBinding(fieldBinding));
+    }
+
+    private static <F> Binding<F> getAccessibleFieldBinding(Binding<?> binding, Field field) {
+        // For final fields
+        field.setAccessible(true);
+
+        Binding<F> fieldBinding = new Binding<F>() {
+            @Override
+            public F get() {
+                try {
+                    return (F) field.get(binding.get());
+                } catch (IllegalAccessException e) {
+                    // Field is public
```

Comment:
Shouldn't this say "private" instead?

Reply:
I meant to convey that the catch block will never be called because the field is public. I'll clarify that further.